tests/quality_pragmaitc/test_refinement.py

```diff
@@ -24,63 +32,430 @@ def test_shouldBeApplicable():
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    stricter = QualityConstraint(
+        anotherContext, CommonMetrics.SECONDS, 10, Comparison.LESS_OR_EQUAL_TO)
+
+    goal.addDependency(task)
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(context)
+    goal.getInterpretation().addQualityConstraint(qc)
+    goal.getInterpretation().addQualityConstraint(stricter)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    fullContext.append(context)
+    # assertEquals(null, goal.isAchievable(fullContext, interp))
+
+    assert qc in goal.getInterpretation().getQualityConstraints(
+        fullContext)
+
+    fullContext.append(anotherContext)
+    assert qc in goal.getInterpretation().getQualityConstraints(
+        fullContext)
+    assert stricter in goal.getInterpretation().getQualityConstraints(
+        fullContext)
+
+    fullContext.pop(context)
+    assert qc not in goal.getInterpretation().getQualityConstraints(
+        fullContext)
+
+    assert stricter in goal.getInterpretation().getQualityConstraints(
+        fullContext)
+
+
+def shouldThereBeMoreThanOneApplicableQCreturnTheStricterOne():
+    goal = Pragmatic(Decomposition.AND)
+
+    task = Task()
+    context = Context("C1")
+    anotherContext = Context("C2")
+
+    fullContext = []
+
+    qc = QualityConstraint(context, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    stricter = QualityConstraint(
+        anotherContext, CommonMetrics.SECONDS, 10, Comparison.LESS_OR_EQUAL_TO)
 
-        set.append(context1)
-        set.append(context2)
+    goal.addDependency(task)
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(context)
+    goal.getInterpretation().addQualityConstraint(qc)
+    goal.getInterpretation().addQualityConstraint(stricter)
+
+    assert stricter == qc.stricterQC(stricter)
+
+    fullContext.append(context)
+    assert qc in goal.getInterpretation().getQualityConstraints(fullContext)
+
+    fullContext.append(anotherContext)
+    assert stricter in goal.getInterpretation().getQualityConstraints(fullContext)
+
+
+def shouldIncludeNonApplicableContexts():
+    goal = Pragmatic(False)
+
+    task = Task()
+    context = Context("C1")
+    wrongContext = Context("C2")
+    current = []
+
+    qc = QualityConstraint(context, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+
+    task.addApplicableContext(context)
+    task.setProvidedQuality(context, CommonMetrics.SECONDS, 13)
+
+    goal.addDependency(task)
+    goal.setIdentifier("Root")
+    #goal.addNonApplicableContext(wrongContext)
+    goal.getInterpretation().addQualityConstraint(qc)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    current.append(wrongContext)
+    #assert goal.isAchievable(current, interp) is None
+
+    current.append(context)
+    assert goal.isAchievable(current, interp) is None
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=4861978)